Initial draft of prerendering-state.md (#11)
jeremyroman/alternate-loading-modes: prerendering-state.md

specification](https://github.com/w3c/page-visibility/issues/42).

With the re-introduction of prererendering, we need to bring back some form of this API.

## Why not `visibilityState=='prerender'`?

The most straightforward way to do this would be to bring back the previously specified `document.visibilityState ==
'prerender'`.  Doing so has a number of drawbacks:

### Not mutually-exclusive

In actuality, some of the above use cases are interested in visibility rather than prerendering state.  For example, an
ordinary (i.e. non-prerendering) page that's hidden would likely want to "avoid running JavaScript animations" and
"avoid fetching large video resources". Thus, pages would have to consider both the `hidden` state as well as a
re-introduced `prerender` to get the desired behavior.

However, the [portals proposal](https://github.com/WICG/portals/) means that `prerender` isn't mutually exclusive
with `visible` either. In that case, we do want to "run JavaScript animations" and "fetch large resources" since it'll
be visible to the user, despite being in a prerendering browsing context (and all the restrictions that come with that).

### Web-compatibility

> Related to the above point, a common pattern on the web is to assume only two visibility states. For example, code such

Reviewer: Hm, this is rather unfortunate since `prerender` was only removed from the spec in 2019 and WebKit just removed it from trunk a few weeks ago in https://bugs.webkit.org/show_bug.cgi?id=215851. It's unfortunate that visibilityState is not a boolean, but we don't want to add new types because sites are treating it as one. But the argument above about prerendering not being mutually exclusive with visibility and hidden makes sense.

We propose adding a new state variable:

```webidl
partial interface Document {
    readonly attribute boolean    prerendering;
    attribute EventHandler        onprerenderingchange;
};
```

`document.prerendering` returns true if the document is currently hosted in a [prerendering browsing
context](browsing-context.md). The `prerenderingchange` event is fired whenever this value changes.

A state separate from visibility ensures existing visibility-based states are correctly accounted for. It also forces
authors to consider their page's behavior in light of prerendering restrictions and non-interactivity and the fact that
a user may not have initiated the page load in the first place.

### Relationship to `visibilityState`

`prerendering` and `visibilityState` are entirely independent and all combinations of states are possible:

| State          | `prerendering` | `visibilityState` |
| -------------- | -------------- | ----------------- |
| Foreground Tab | false          | 'visible'         |
| Background Tab | false          | 'hidden'          |
| Portal         | true           | 'visible'         |
| Prerender      | true           | 'hidden'          |

> <<<<<<< HEAD

Reviewer: remove?

# Prerendering State API

## Summary

We propose introducing a boolean property `document.prerendering` and associated change event to distinguish
prerendering browsing contexts from regular ones.

_Note: This is based on the more general discussion in this [background

> doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit?usp=sharing)_

Reviewer 2: ```suggestion
doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit?usp=sharing)._
```

## Background

Pages need to know when they're being rendered inside a [prerendering browsing context](browsing-context.md).
Some examples of divergent behavior:

* Avoid fetching large video resources
* Avoid measuring ad impressions

> * Avoid running javascript animations

Reviewer 2: ```suggestion
* Avoid running JavaScript animations
```

* Avoid measuring user interactions for analytics
* Avoid starting server-coordinated behavior (e.g. ticket purchase countdown).

> A previous iteration of prerendering used [page-visibility](https://www.w3.org/TR/page-visibility-2) to denote a

Reviewer 2: ```suggestion
A previous iteration of prerendering used [page-visibility](https://w3c.github.io/page-visibility/) to denote a
```

prerendering context. However, this was unimplemented and [removed from the
specification](https://github.com/w3c/page-visibility/issues/42).

With the re-introduction of prererendering, we need to bring back some form of this API.

> ## Why not visibilityState=='prerender'?

Reviewer 2: ```suggestion
## Why not `visibilityState=='prerender'`?
```

> The most straightforward way to do this would be to bring back the previously specified `visibilityState ==

Reviewer 2: ```suggestion
The most straightforward way to do this would be to bring back the previously specified `document.visibilityState ==
```

'prerender'`.  Doing so has a number of drawbacks:

### Not mutually-exclusive

In actuality, some of the above use cases are interested in visibility rather than prerendering state.  For example, an

> ordinary (i.e. non-prerendering) page that's hidden would likely want to "avoid running javascript animations" and

Reviewer 2: ```suggestion
ordinary (i.e. non-prerendering) page that's hidden would likely want to "avoid running JavaScript animations" and
```

"avoid fetching large video resources". Thus, pages would have to consider both the `hidden` state as well as a
re-introduced `prerender` to get the desired behavior.

However, the [portals proposal](https://github.com/WICG/portals/) means that `prerender` isn't mutually exclusive

> with `visible` either. In that case, we do want to "run javascript animations" and "fetch large resources" since it'll

Reviewer 2: ```suggestion
with `visible` either. In that case, we do want to "run JavaScript animations" and "fetch large resources" since it'll
```

be visible to the user, despite being in a prerendering browsing context (and all the restrictions that come with that).

### Web-compatibility

Related to the above point, a common pattern on the web is to assume only two visibility states. For example, code such
as:

> ```

Reviewer 2: ```suggestion
```js
```

onload = () => {
  if (document.visibilityState == 'visible')
    // Load large images
  else
    // Defer loading large images
}
```

An analysis [found several
examples](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit#heading=h.rkorueof7xev)
of such behavior.

(re-)Introducing a `prerender` state could break such pages. In particular, portals (or some other, future

> prerendering-while-visible context) would show static or missing content. A hidden prerender may not have immediate

Reviewer 3: The term "preview" might be a more intuitive way of describing prerendering-while-visible.

Author: I agree, but I think it makes sense for this be a signal that we're in a prerendering-browsing-context. If that's the case, I think it'd make sense to match the names of the two things;  "prerendering-browsing-context" has the same issue. Would it make sense to rename them both? @jeremyroman @domenic for their thoughts.

Reviewer 3: Ah, I was only suggesting we might use the term preview in the prose describing what the new prerender state did, not changing the name of anything web-facing.

Author: Ah - true, but that does beg the question whether naming the web-facing APIs 'prerender' evokes the right meaning if we need to clarify in prose

Author: This is still outstanding but unresolved. I filed #12 to keep this in mind but I think we can punt on this for now and leave it as "prerendering" since it's much clearer to outsiders what we're talking about. As other use cases shape up we can revisit it in the future.

user-visible effects, but may use more resources than necessary.

> ## document.prerendering

Reviewer 2: ```suggestion
## `document.prerendering`
```

We propose adding a new state variable:

> ```

Reviewer 2: ```suggestion
```webidl
```

partial interface Document {
    readonly attribute boolean    prerendering;
    attribute EventHandler        onprerenderingchange;
};
```

`document.prerendering` returns true if the document is currently hosted in a [prerendering browsing
context](browsing-context.md). The `prerenderingchange` event is fired whenever this value changes.

A state separate from visibility ensures existing visibility-based states are correctly accounted for. It also forces
authors to consider their page's behavior in light of prerendering restrictions and non-interactivity and the fact that
a user may not have initiated the page load in the first place.

> ### Relationship to visibilityState

Reviewer 2: ```suggestion
### Relationship to `visibilityState`
```

`prerendering` and `visibilityState` are entirely independent and all combinations of states are possible:

| State          | `prerendering` | `visibilityState` |
| -------------- | -------------- | ----------------- |
| Foreground Tab | false          | 'visible'         |
| Background Tab | false          | 'hidden'          |
| Portal         | true           | 'visible'         |
| Prerender      | true           | 'hidden'          |

> #### VisibilityState as a termination signal

Reviewer 2: ```suggestion
#### `visibilityState` as a termination signal
```

Authors have been [encouraged](https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/) to
switch from `unload` and `beforeunload` events to `visibilityState=='hidden'` as a signal that their page may be

> terminated. See also [PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) for related discussion.

Reviewer 2: ```suggestion
terminated. See also [w3c/PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) for related discussion.
```

> Entering a `prerendering` state, via being put into a portal, is likely a point where such a termination signal should

Reviewer 2: maybe "being put into a portal (via portal predecessor adoption)"?

Author: Done

be fired; it is analogous to navigating to a new page. However, this means pages would now also have to listen for a
switch to `prerendering == true`.

This [background
doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit#heading=h.acmnp6zdmcik) deals
extensively with this issue. An ideal solution would introduce a new event for this use case, rather than implicitly

> tying it to any particular state.

Reviewer 2: Maybe note that you think that can be done as a followup, and in the meantime pages won't handle the case of being put into a portal perfectly, and that's OK?

Author: Done

> ### Relationship to loading-modes

Reviewer 2: Let's delete this section and roll it into the TODO you inserted above, which I commented on.

Author: Done

[Alternate-loading-modes](https://github.com/bokand/alternate-loading-modes/blob/main/browsing-context.md#current-proposal)
proposes explicit APIs to expose the exact loading mode used. Even with `document.prerendering`, this additional
information may still be useful for specialized cases where fetching may need to depend on the exact loading mode (TODO:
examples?).

However, we believe `document.prerendering` will cover the most common use cases without getting into the level of
granularity of fetching restrictions.

Note: it's possible that `document.prerendering` may be  just syntactic sugar for some subset of
`document.loadingMode.type`, e.g. `document.loadingMode.type != 'default'` or `document.loadingMode.type ==
'prerender' || document.loadingMode.type == 'uncredentialed-prerender'`.

> ### Relationship to other states

Reviewer 2: I wonder if you want an "open questions" h2-level section, which contains "visibilityState as a termination signal", "relationship to other states", and "Naming bikeshed"?

Author: Good idea, done.

As noted in the [background
doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit#heading=h.14z99pd6akf0), there
are other states where it may make sense to report `prerendering` state. For example, a page loading into a mobile
background tab maybe terminated, without notice, before the user ever sees it. Similarly, the live-preview in a mobile app
or tab switcher is a non-interactive preview, reminiscent of a portal.

Should these states also report `prerendering==true`? Should they use the prerendering browsing context concept? The
answers here aren't clear but this is worth considering.

> ### Naming bikeshed

Reviewer 2: Good summary of the issues!

Author: Thanks :)

We've called this state `prerendering` here to make clear the association between it and a [prerendering browsing
context](browsing-context.md). However, this _may_ be misleading in some cases. Portals aren't exactly prerendering in the way most authors
would think about it. Additionally, a portal may enter the prerendering state from a regular context; this feels
unintuitive to call "prerendering".

We should also consider future use cases. For example, if
[fenced-frames](https://github.com/shivanigithub/fenced-frame/) were to use this special browsing context, does it make
sense to call it "prerendering"?

The term "interactive" seems appealing as it captures a major difference between prerendering/portals and regular
browsing contexts. However, it would be unintuitive to explain how a background tab is considered "interactive". In

> particular, if we want this to solve [PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) we'd have the

Reviewer 2: ```suggestion
particular, if we want this to solve [w3c/PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) we'd have the
```

bizarre case that entering the app switcher would make a page non-interactive, then backgrounding it from there would make

> it interactive.

Reviewer 4: It might be worth mentioning how it does (or doesn't, IIUC) interact with bfcache.

Author: I've been thinking of `document.prerendering` as "is in prerendering browsing context". Does BFCache have any expected interaction with the concept of a prerendering browsing context introduced in this repo? I expect not (and in that case I'll add that here) but I'm not sure if we're thinking of reusing some of the rendering/permissions restrictions.

Reviewer 4: I think likely not. Though it's similar in some regards to prerendering, script can't observe that much because it gets stopped for the duration where that would apply. But I think it's non-obvious enough to note.

Reviewer: +1 could be good to mention this. My understanding also is that bfcached pages do not run script so can't observe document.prerendering.
